### PR TITLE
[stable/prometheus-operator] updated to 0.31.1

### PR DESCRIPTION
**Pull request by caarlos0**
#### What this PR does / why we need it:
- updates to latest prometheus-operator and prometheus
- updates its CRDs
- ~~updates rules (used the script on the `hack` folder for that)~~ will do in another PR

#### Special notes for your reviewer:

Would be nice to have a script to generate the updated CRD files... I can happily work on that as well :)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

### CODE REVIEW

**Comment by runningman84** [comment]
What is the difference between a servicemonitor and a pod monitor?

**Comment by caarlos0** [comment]
> What is the difference between a servicemonitor and a pod monitor?

according to the [changelog](https://github.com/coreos/prometheus-operator/releases/tag/v0.31.0):

> "allowing monitoring pods directly without the necessity to go through a Endpoints of a Service, this is an experimental feature, it may break at any time without notice."

**Comment by vsliouniaev** [comment]
This doesn't create the CRD when I do an upgrade. 

I realize that the other CRDs have this if statement around them, but that's a workaround for an issue with upfrading CRDs that already existed with the various helm install hooks.

I haven't experimented enough with this yet to understand the correct approach, but I really think that we need to solve this issue so that we have the flexibility to use the CRDs as part of the chart. 

/hold

**Comment by caarlos0** [comment]
@vsliouniaev yeah I noticed that too... installed the CRD manually as stated on the README ("Helm fails to create CRDs")... not sure how to properly fix this either, and the worst part is that without the CRD the operator enters in crashloopbackoff evenually... 

anyway, let me know how you want me to proceed.

**Comment by caarlos0** [comment]
any ideas for this?

File: stable/prometheus-operator/ci/test-values.yaml
Comment (vsliouniaev) [inline stable/prometheus-operator/ci/test-values.yaml]: `0.31.1` has been released, would you mind updating this here?

File: stable/prometheus-operator/Chart.yaml
Comment (vsliouniaev) [inline stable/prometheus-operator/Chart.yaml]: 0.31.1 has been released, would you mind updating this here?

**Review by vsliouniaev — CHANGES_REQUESTED** [review]
After doing some testing, I think the changes that we need to do are:
- Remove the `.Release.IsInstall` from all CRD templates
- Remove "helm.sh/hook-delete-policy": "before-hook-creation" from all CRD templates

this appears to work. I suggest you try do some tests upgrading the chart from an earlier stable version to this one and see if you run into any issues. This appears to be working fine in my case on `kind` back with even Helm 2.11.0

**Comment by mikesplain** [comment]
I did similar testing to @vsliouniaev suggested and it worked as well. Looking forward to getting this in!

File: stable/prometheus-operator/ci/test-values.yaml
Comment (caarlos0) [reply stable/prometheus-operator/ci/test-values.yaml]: sure np

**Comment by caarlos0** [comment]
tried on k3s, works as expected!

**Comment by caarlos0** [comment]
BTW I'm being using this a lot, so if you folks need help on the prs/issues etc I'm glad to add myself as a reviewer/approver in another PR

**Review by gzamboni — APPROVED** [review]

**Comment by caarlos0** [comment]
couldn't reproduce this locally 🤔

/retest

File: stable/prometheus-operator/templates/prometheus/rules/k8s.rules.yaml
Comment (vsliouniaev) [inline stable/prometheus-operator/templates/prometheus/rules/k8s.rules.yaml]: Did you intend to include a rules change in this PR? I re-ran the dashboards and rules sync scripts and they look like they're pulling in quite  a lot more changes than this currently.

There have been issues syncing rules and dashboards in the past, so I think it's worth going through them in detail when they're done. Do you want to make a separate PR for that or do you want to re-run the full sync scripts and validate the result?

Comment (caarlos0) [reply stable/prometheus-operator/templates/prometheus/rules/k8s.rules.yaml]: I'll undo this change then... its safer I guess :)

**Comment by vsliouniaev** [comment]
I think the one last thing is that the readme values are now out of sync with the info in `values.yaml`, could you update these? I'll merge the PR once that's in. Apologies, for the wait, again.

**Comment by caarlos0** [comment]
> I think the one last thing is that the readme values are now out of sync with the info in `values.yaml`, could you update these? I'll merge the PR once that's in. 

done, my bad I forgot to change it. 

> Apologies, for the wait, again.

no problem at al! thanks for keeping on top of this :)

**Comment by vsliouniaev** [comment]
```
I0627 12:27:54.908] Error: customresourcedefinitions.apiextensions.k8s.io "alertmanagers.monitoring.coreos.com" already exists
```

This is one of the reasons that we had the pre-install hook to remove any in-cluster definitions before the chart was provisioned. I think that the way the tests are executed in the CI system can cause these kinds of things to happen. I think it's a good change though, so I'd rather have a worse CI experience if it improves the CRD experience.

/retest

**Comment by caarlos0** [comment]
> This is one of the reasons that we had the pre-install hook to remove any in-cluster definitions before the chart was provisioned. I think that the way the tests are executed in the CI system can cause these kinds of things to happen. I think it's a good change though, so I'd rather have a worse CI experience if it improves the CRD experience.

I guess we can't delete the CRDs before installing on CI? :(

**Comment by vsliouniaev** [comment]
The crd-install hook is the very first thing that is executed by helm. I actually think that if we have a different PR delete them, then this PR will pass.

**Comment by vsliouniaev** [comment]
How about we add a CI-only condition to run the pre-install delete hook for the CRDs? That should work I think.

**Comment by caarlos0** [comment]
> How about we add a CI-only condition to run the pre-install delete hook for the CRDs? That should work I think.

makes sense, I don't know how to do that though... tried to search code for `isci`, `is_ci` and etc didn't find anything.. if you can direct me on how to do it I'll add it to this PR )

**Comment by vsliouniaev** [comment]
The CI process uses the values file from the ci folder. The docs for this are somewhere in the root of this repo. So you can add something like `useDeleteBeforeCreationHooks` and only enable it in the ci values.

**Comment by caarlos0** [comment]
@vsliouniaev something like this?

**Comment by vsliouniaev** [comment]
First thing that runs is the crd-install hook.
You just want to add these around these lines: https://github.com/helm/charts/pull/14943/files#diff-f82a702183a715ecb635f41469329b90L13

So that they only happen during ci and not when normal people are using the chart

**Comment by caarlos0** [comment]
I'm sorry, I didn't understand what you mean... can you elaborate?

(probably my fault because I'm not experienced with these helm hooks stuff)

**Comment by vsliouniaev** [comment]
I've tried to implement my suggestion but it does not work, so there isn't a point in creating this extra value. I'm trying to look at alternatives.
I think the cause itsn't this PR, but rather some other PR that put the CI cluster in this state. Could you rever that last change and I'll try to get the build to pass afterward?

**Comment by vsliouniaev** [comment]
No worries!

I don't think there's a lot that can be done to actually fix the issue - there are many people having issues with CRDs and Helm and the CI doesn't run on a clean cluster each time

Suggest you make this change: https://github.com/helm/charts/commit/1ce6f076f4fce51a704974b8c20bd088f5e02aef

**Comment by caarlos0** [comment]
awesome! will cherry pick that, thanks!

**Comment by caarlos0** [comment]
awesome, thanks!

**Pull request merged**
